### PR TITLE
Task/selenium tests

### PR DESCRIPTION
**Pull request by birdsarah**
@mistakevin - just getting started on this with a TapTool and Canteen.

It works great with the FF driver, when I ran it with PhantomJS driver, it had a different hash. Are the Canteen hashes different for different browsers?

### CODE REVIEW

**Comment by birdsarah** [comment]
Also @mistakevin, does the travis failure mean anything to you?

**Comment by mattpap** [comment]
@birdsarah, looking at the log, `canteen` wasn't installed as expected.

**Comment by mistakevin** [comment]
This is interesting. It appears that we are getting different values between Chrome and Firefox :grimacing: 
screen shot 2015-08-17 at 11 31 09 am
@birdsarah @bryevdv @mattpap

File: scripts/travis_install [hunk 1]
```diff
@@ -40,6 +40,7 @@ conda install --yes abstract-rendering
 conda install --yes multiuserblazeserver pillow
 ##hack - we need latest blaze right now
 conda install --yes -c blaze blaze
+pip install git+https://github.com/birdsarah/pytest-selenium.git
```
Comment (mattpap) [inline scripts/travis_install]: This should get a conda package in `bokeh` channel at some point.

Comment (birdsarah) [inline scripts/travis_install]: Yes indeed. I want to use the maintained package not my fork, and I think I just got the answer to make that possible.  Hopefully @damianavila will be able to help get it onto bokeh channel then.

**Comment by birdsarah** [comment]
@mattpap - yep - installed manually locally works for me - @mistakevin has filed a report on this with npm https://github.com/npm/npm/issues/9254 - I'm not sure what to do about it though - I suppose we could just put `npm install platfora/Canteen` in the `travis_install` for now?

**Comment by mattpap** [comment]
>  I suppose we could just put npm install `platfora/Canteen` in the `travis_install` for now?

If we can't figure out anything better, then temporarily yes. Just make sure to update documentation, because it's a burden not only on travis-ci, but also for anyone who would like to use canteen-based infrastructure locally.

**Comment by mattpap** [comment]
> It appears that we are getting different values between Chrome and Firefox.

Which example this refers to?

**Comment by mistakevin** [comment]
@birdsarah `npm install` has started working for me. I'm not sure what changed, but give it a try locally to see if you are seeing the same behavior.

@mattpap the example is `examples/charts/file/boxplot.py`, with on my `mistakevin/feature/canteen` branch

**Comment by birdsarah** [comment]
@mattpap @mistakevin - in terms of the canteen hashes changing. Messing around with any of the examples once you have canteen installed (and a new version of BokehJS compiled should work). In this PR, I was specifically referring to the following:

- `py.test -m integration` (which will run with Firefox as default)
- this should pass

Then try:

- `py.test -m integration --driver=PhantomJS` or `py.test -m integration --driver=Chrome`
- these doesn't pass as the hashes are different

(note that you need selenium ChromeDriver on your path for Chrome to work).

**Comment by mattpap** [comment]
> `npm install` has started working for me

I don't see any improvement (speaking about version 2.13.5).

**Comment by birdsarah** [comment]
> I don't see any improvement (speaking about version 2.13.5).

I am also still having this problem. I started using `nvm` as @mistakevin recommended. I am using `nvm use stable` this gives me npm v2.11.3.

**Comment by mattpap** [comment]
@birdsarah, use `npm install -g npm` (perhaps with `sudo`) to get the most recent version.

**Comment by damianavila** [comment]
I have built packages for `pytest-selenium` from the **pypi** available version, but they did not work...
So I used the more recent (github) version from Dave Hunt... for that I have to build some other packages (pytest-html and pytest-variables) because the dependencies are different in the new version and there was not available conda packages... **but** we should probably ask the authors to know when will be released the new version and use that... **please**, in the future, I would like to avoid using something not released yet... now there seems to be some real failures in 2726...

**Comment by mistakevin** [comment]
@birdsarah Scouring the internet, there's mention that canvas sub-pixels can render differently in different browsers. Here's an interesting tidbit from the [canvas specs](https://html.spec.whatwg.org/#pixel-manipulation):

     user agents may round the arguments of these methods

and here's a (possibly) relevant [SO](http://stackoverflow.com/questions/12423716/what-is-the-current-state-of-sub-pixel-accuracy-in-the-major-browsers). Not sure what the best solution is here. The `loose:true` might method be too "loose", as it wouldn't tell us if something was way off. I like that we can adjust the `NUMBER_PRECISION`, but until he makes it configurable, we'd have to fork it to test it.

**Comment by birdsarah** [comment]
Just to reference - here's some thoughts from the platfora people: https://github.com/platfora/Canteen/issues/2

**Comment by birdsarah** [comment]
@mistakevin - did you see the latest on the platfora issue - they've implemented a rounding feature in 1.0.4 - should we upgrade?

**Comment by mistakevin** [comment]
Yeah, let's give it a shot. He doesn't version his releases through github, so just deleting it from node_modules and reinstalling it should pull in his latest master (which we should probably lock down in the future with a commit hash)

**Comment by birdsarah** [comment]
@mistakevin - I think this is working for an initial commit of selenium tests. Can you take a look?

It:
* Adds a single selenium test
* That runs on FF (both locally and on travis)
* That selects glyphs on a page and checks the canteen hash

Cheers,

Bird

**Comment by mistakevin** [comment]
@birdsarah What was the final resolution with `Canteen` vs `canteen` with mocha testing?

**Comment by birdsarah** [comment]
@mistakevin - if everything's uppercase and Canteen is in devDependencies then everything's happy - npming, building and testing. I don't know if this is "wrong" in some way - but it does work.

**Comment by mistakevin** [comment]
Have you tried taking it out of the `"browser"` hash altogether?

**Comment by mistakevin** [comment]
LGTM

**Comment by birdsarah** [comment]
ping @mattpap @damianavila @brendancol - would love your review of this PR.

File: .travis.yml [hunk 2]
```diff
@@ -36,6 +36,8 @@ before_install:
     - echo $TRAVIS_BUILD_ID > __travis_build_id__.txt
     - nvm install stable
     - nvm use stable
+    - "export DISPLAY=:99.0"
+    - "sh -e /etc/init.d/xvfb start"
```
Comment (damianavila) [inline .travis.yml]: can you add a comment here for future readers? this is something specific for GUI testing in TravisCI but people is not usually aware of that...

File: bokehjs/package.json [hunk 3]
```diff
@@ -21,6 +21,7 @@
     "browserify": "^6.2.0",
     "browserify-eco": "^0.2.4",
     "browserify-shim": "^3.8.0",
+    "Canteen": "platfora/Canteen",
```
Comment (damianavila) [inline bokehjs/package.json]: so we are using just master? without any hash??? we need to point to an specific commit...

**Comment by damianavila** [comment]
OK, are we pointing to merge this for 0.9.3? We depend on two libraries which are not released, as I said previously, I understand they are things needed to make the work easy and avoid duplication, but we need reference points: for pytest-selenium, we have at least a package referring to an specific commit, but from my understanding (which is low in this matter) we are depending on Canteen master... I don't think this should be merge until we have those specific and "traceable" points...

**Comment by mistakevin** [comment]
We can lock down Canteen to a certain commit with a [commitish-suffix](https://docs.npmjs.com/files/package.json#github-urls), or, to be even safer, we can always fork the repo and use that.

**Comment by damianavila** [comment]
>We can lock down Canteen to a certain commit with a commitish-suffix, or, to be even safer, we can always fork the repo and use that.

I would like to avoid forking if there is no real need for that (in the sense I understand fork in the OS community), so I would be OK with a commitish suffix...

**Comment by birdsarah** [comment]
@damianavila, agree it's good to be pinning these things down. I'm pretty sure that @davehunt will be releasing pytest-selenium soon. In the meantime, given that we are installing it from the bokeh conda repo, is there any need to do anything further. We don't pin any of the other conda packages. There are actually a couple of small tweaks in pytest-selenium coming that I'm excited about, although just around how you call it, nothing in the core functionality.. We can next update our conda package and code when pytest-selenium gets release I think.

**Comment by damianavila** [comment]
>In the meantime, given that we are installing it from the bokeh conda repo, is there any need to do anything further. 

Nop, no need to make anything further... the version I build is actually labelled with the commit hash and we are only installing it from our bokeh user in Anaconda.org.

>There are actually a couple of small tweaks in pytest-selenium coming that I'm excited about, although just around how you call it, nothing in the core functionality.. We can next update our conda package and code when pytest-selenium gets release I think.

Yep, I agree...

**Comment by damianavila** [comment]
OK with the changes added, this LGTM, I will merge it now...

**Pull request merged**